plugin/environment.go

```diff
@@ -162,6 +164,27 @@ func (env *Environment) Statuses() (model.PluginStatuses, error) {
 	return pluginStatuses, nil
 }
 
+// Manifest returns a manifest for a given pluginId.
+// Returns ErrPluginNotFound if plugin is not found.
+func (env *Environment) Manifest(pluginId string) (*model.Manifest, error) {
+	plugins, err := env.Available()
+	if err != nil {
+		return nil, errors.Wrap(err, "unable to get plugin statuses")
+	}
+
+	for _, plugin := range plugins {
+		if plugin.Manifest == nil {
+			continue
+		}
+
+		if plugin.Manifest.Id == pluginId {
+			return plugin.Manifest, nil
+		}
+	}
+
+	return nil, ErrPluginNotFound
```

Comment:
Ah, I didn't realize this was a new method -- I'd propose `return nil, nil` for `not found` as the idiomatic Go approach.

Reply:
@lieut-data 
I may be wrong, but I thought it was Go idiomatic to return `return nil, ERR` when an entity is not found. And return `return nil, nil` for lists e.g. `([]Plugin, error)`. But I couldn't find any solid guidelines in https://golang.org/doc/effective_go.html.

I do see similar patterns set up for store-like APIs, e.g. https://golang.org/pkg/database/sql/#DB.QueryRow

Albeit this may be a dated example: https://github.com/golang/go/blob/master/src/os/exec/lp_windows.go#L92

Reply:
Perhaps my claim to this being idiomatic was too strong: I too can't find a convincing proof one way or another. Accessing a map won't fail with an `err`, so it's not a convincing argument, but I've aways appreciated the zero-value return in that case vs. the need to expose an error to report that the object requested doesn't exist.

Part of my thinking has developing from seeing awkward cases like this:
https://github.com/mattermost/mattermost-server/blob/0ddc5f4a3245cee4db737da0f3e8772f5b92e232/app/team.go#L918-L924

Admittedly, this is a poor example given the code is returning a pointer to a slice of pointers. But it illustrates my thinking: if it's natural to expect the zero value, then it shouldn't be awkward to handle this case.

In your case, I'd argue that it's natural to expect that a manifest might not exist. Actually, maybe there's a better way to think about this:
https://github.com/mattermost/mattermost-server/blob/3bd975c1bc7e45129f76e5a1e110f40a3dec8e37/app/plugin.go#L431-L434

In the above code, if any /other/ error occurs, we assume the plugin is just not installed and otherwise march on, but that might obscure some other problem. If we changed the code to handle this:
```go
marketplacePlugin := &model.MarketplacePlugin{
	BaseMarketplacePlugin: plugin,
})

if manifest, err = pluginsEnvironment.GetManifest(plugin.Manifest.Id); err == nil {
    // Plugin is installed.
	marketplacePlugin.InstalledVersion = manifest.Version		
} else if err != nil && err != plugin.ErrPluginNotFound {
    return model.NewAppError("GetMarketplacePlugins", "app.plugin.config.app_error", nil, err.Error(), http.StatusInternalServerError)
}
```

I'd argue that the special handling of the error is unnecessary, and would be more clearly expressed as:
```go
if manifest, err = pluginsEnvironment.GetManifest(plugin.Manifest.Id); err != nil {
    return model.NewAppError("GetMarketplacePlugins", "app.plugin.config.app_error", nil, err.Error(), http.StatusInternalServerError)
} else if manifest != nil {
    // Plugin is installed.
	marketplacePlugin.InstalledVersion = manifest.Version		
}
```

Reply:
@lieut-data 

> Perhaps my claim to this being idiomatic was too strong: I too can't find a convincing proof one way or another. Accessing a map won't fail with an `err`, so it's not a convincing argument, but I've aways appreciated the zero-value return in that case vs. the need to expose an error to report that the object requested doesn't exist.
> 

Ah, but I don't think the comparison is fair though. A two-value assignment does sort of return an `err` equivalent `i, ok := m["route"]`, no? :)

But I do see the awkwardness in the examples that you shared, however I would argue the messiness can come from the way the code is organized (apart from handling `not found` as a special case). This does seem to be a common pattern in the stdlib though https://golang.org/pkg/os/#IsNotExist.

I would've expressed the example as:
```
	channelList := &model.ChannelList{}
	if channelList, err = a.Srv.Store.Channel().GetChannels(team.Id, user.Id, true); err != nil && err.Id != "store.sql_channel.get_channels.not_found.app_error" {
		return err
	}
```

I think it all comes down to our convention we set here. I'm personally in the favour of returning `err` (not found) for single entity APIs. However, I'd be OK if we decided to use `return nil, nil` instead, and all APIs moving forward. Overall 2/5 on this.

Reply:
Actually everywhere I can remember `not_found_error` is returned explicitly. For me `nil, nil` is bit confusing. I'd stick to `ErrPluginNotFound`.

Reply:
> Ah, but I don't think the comparison is fair though. A two-value assignment does sort of return an err equivalent i, ok := m["route"], no? :)

IMO, what's special about `ok`, here, is that everything I need to interpret it is immediately available: I don't need to compare `ok` to a sentinel value exported by the package in order to interpret it. Similarly, returning `nil` for a pointer without an error -- to me, at least -- signals that that object didn't exist.

However, I can see how this is subjective on my part, and I also see how this approach is incompatible with just returning values instead of pointers: something else we've agreed is overused at present. Lastly, I think the `sql.ErrNoRows` paradigm settles the matter as an idiomatic example of why returning an error for not found makes sense.

Can you fix this code to handle non-`plugin.ErrPluginNotFound` errors?
https://github.com/mattermost/mattermost-server/blob/3bd975c1bc7e45129f76e5a1e110f40a3dec8e37/app/plugin.go#L431-L434

Lastly, to firm up the proposal going forward, is there one of these sentinel values per package? Multiple for disparate kinds of objects? Implement a common `NotFound` interface? (I'm fine with deferring all of these questions, but just throwing it out for discussion.)

Reply:
@lieut-data sounds good.

Went with a generic `plugin.ErrNotFound` for now to cover most cases, and can add entity specific errors in the future if need be.